libs/ml/include/ml/optimisation/lazy_adam_optimiser.hpp

```diff
@@ -0,0 +1,371 @@
+#pragma once
+//------------------------------------------------------------------------------
+//
+//   Copyright 2018-2019 Fetch.AI Limited
+//
+//   Licensed under the Apache License, Version 2.0 (the "License");
+//   you may not use this file except in compliance with the License.
+//   You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+//   Unless required by applicable law or agreed to in writing, software
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+//
+//------------------------------------------------------------------------------
+
+#include "math/standard_functions/pow.hpp"
+#include "math/standard_functions/sqrt.hpp"
+#include "ml/core/graph.hpp"
+#include "ml/optimisation/optimiser.hpp"
+
+#include <memory>
+#include <string>
+#include <vector>
+
+namespace fetch {
+namespace ml {
+namespace optimisers {
+
+/**
+ * LazyAdam is a variant of the Adam optimizer that handles sparse updates more efficiently.
+ * The original Adam algorithm maintains moving-average accumulators for each trainable variable;
+ * the accumulators are updated at every step. This class provides lazier handling of gradient
+ * updates for sparse variables. It only updates moving-average accumulators for sparse variable
+ * indices that appear in the current batch, rather than updating the accumulators for all indices.
+ * Compared with the original Adam optimizer, it can provide large improvements in model training
+ * throughput for some applications. However, it provides slightly different semantics than the
+ * original Adam algorithm, and may lead to different empirical results.
+ * @tparam T TensorType
+ */
+template <class T>
+class LazyAdamOptimiser : public Optimiser<T>
```

Comment:
doesn't it make more sense to inherit from AdamOptimiser and then just override the relevant methods?